From 3b31f2b3b106b717f504bed48c866eab4ec0b2f1 Mon Sep 17 00:00:00 2001
Message-Id: <3b31f2b3b106b717f504bed48c866eab4ec0b2f1.1476965418.git.keller@fs-net.de>
In-Reply-To: <1bbe4f4c985819c8f8ead54288ca1cc8a4b7f51f.1476965418.git.keller@fs-net.de>
References: <1bbe4f4c985819c8f8ead54288ca1cc8a4b7f51f.1476965418.git.keller@fs-net.de>
From: Hartmut Keller <keller@fs-net.de>
Date: Wed, 19 Oct 2016 19:05:04 +0200
Subject: [PATCH 11/12] Optimize buffer copy in unaligned case

In gal2d-renderer, even if the buffer needs to be copied from a buffer that
is not aligned for g2d access to an aligned buffer, we can copy whole
pixel rows. We do not need to copy every pixel on its own.
---
 src/gal2d-renderer.c |   12 ++++--------
 1 files changed, 4 insertions(+), 8 deletions(-)

diff --git a/src/gal2d-renderer.c b/src/gal2d-renderer.c
index 73559e7..da0f701 100644
--- a/src/gal2d-renderer.c
+++ b/src/gal2d-renderer.c
@@ -409,16 +409,12 @@ gal2dBindBuffer(struct weston_surface* es)
 		}
 		else
 		{
-			int i, j;
+			int i;
+
 			for (i = 0; i < buffer->height; i++)
 			{
-				for (j = 0; j < buffer->width; j++)
-				{
-					gctUINT dstOff = i * alignedWidth + j;
-					gctUINT srcOff = (i * buffer->width + j);
-
-					memcpy(va + dstOff * 4, logical + srcOff * 4, 4);
-				}
+				memcpy(va + i * alignedWidth * 4,
+					   logical + i * buffer->width * 4, buffer->width * 4);
 			}
 		}
 		gcmVERIFY_OK(gcoSURF_Unlock(surface, (gctPOINTER)va));
-- 
1.7.4.4

